### PR TITLE
sampled_addmm: backward performance improvements

### PR DESCRIPTION
**Pull request by nikitaved**
No need to do double `sparse_mask`, let's squash everything into one call!
This PR exercises https://github.com/pytorch/pytorch/pull/103750, so here is an autogened code for the backward pass.

```
at::Tensor sparse_sampled_addmm(c10::DispatchKeySet ks, const at::Tensor & self, const at::Tensor & mat1, const at::Tensor & mat2, const at::Scalar & beta, const at::Scalar & alpha) {
  auto& self_ = unpack(self, "self", 0);
  auto& mat1_ = unpack(mat1, "mat1", 1);
  auto& mat2_ = unpack(mat2, "mat2", 2);
  [[maybe_unused]] auto _any_requires_grad = compute_requires_grad( self, mat1, mat2 );

  std::shared_ptr<SparseSampledAddmmBackward0> grad_fn;
  if (_any_requires_grad) {
    grad_fn = std::shared_ptr<SparseSampledAddmmBackward0>(new SparseSampledAddmmBackward0(), deleteNode);
    grad_fn->set_next_edges(collect_next_edges( self, mat1, mat2 ));
    grad_fn->alpha = alpha;
    grad_fn->beta = beta;
    if (grad_fn->should_compute_output(2)) {
      grad_fn->mat1_ = SavedVariable(mat1, false);
    }
    if (grad_fn->should_compute_output(1)) {
      grad_fn->mat2_ = SavedVariable(mat2, false);
    }
    grad_fn->self_ = SavedVariable(self, false);
  }

```

As you can see, we do not save tensors unless needed.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #101978
* #103548
* #104042
* __->__ #103544
* #103750

cc @alexsamardzic @pearu @cpuhrsch @amjames @bhosmer @ezyang @albanD @zou3519 @gqchen @soulitzer @Lezcano @Varal7

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
I'm not sure we want to do that because this will regress memory usage :/ 
In particular, if only self requires grad, m1 and m2 won't be saved today. But with this new code, they will always be saved.

I think there was a discussion on fixing this, do you remember where it is @soulitzer ?
Maybe something like `sparse_sampled_addmm_backward(grad, self, optional_save_if(mat2.requires_grad(), mat1), optional_save_if(mat1.requires_grad(), mat2), alpha, beta, grad_input_mask)` and make the backward take optional<> args.
We then need to use our pattern matcher for replacement to make a smart decision on saving.

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -2441,9 +2441,7 @@
   result: replication_pad3d_backward_symint(grad_output_t, self_p, padding)
 
 - name: sparse_sampled_addmm(Tensor self, Tensor mat1, Tensor mat2, *, Scalar beta=1, Scalar alpha=1) -> Tensor
-  self: maybe_multiply(grad, beta.conj())
-  mat1: maybe_multiply(grad.sparse_mask(self).mm(mat2.mH()), alpha.conj())
-  mat2: maybe_multiply(mat1.mH().mm(grad.sparse_mask(self)), alpha.conj())
+  self, mat1, mat2: sparse_sampled_addmm_backward(grad, self, mat1, mat2, alpha, beta, grad_input_mask)
```
Comment (nikitaved) [inline tools/autograd/derivatives.yaml:2444]: To avoid any confusion, the gradient wrt `self` is incorrect. It is fixed in https://github.com/pytorch/pytorch/pull/103548.

**Comment by soulitzer** [comment]
Here's the issue about multi-output functions saving unnecessary tensors https://github.com/pytorch/pytorch/issues/97575, it should be a small codegen fix

**Comment by nikitaved** [comment]
@albanD , @soulitzer , would the current code be sufficient for now? Or is there a way to tell autogen to produce and reuse some intermediaries?

**Comment by soulitzer** [comment]
There's no way to tell the codegen to do this currently unfortunately

There needs to be an API to specify what needs to be saved for what, and then some code gen updates to translate that information into extra logic in the VariableType kernel

And what alban is proposal here could be promising:
> Maybe something like sparse_sampled_addmm_backward(grad, self, optional_save_if(mat2.requires_grad(), mat1), optional_save_if(mat1.requires_grad(), mat2), alpha, beta, grad_input_mask) and make the backward take optional<> args.
We then need to use our pattern matcher for replacement to make a smart decision on saving.

**Comment by nikitaved** [comment]
What @albanD suggested is already applied here :)

**Comment by soulitzer** [comment]
So if we land this PR as-is we're just trading off memory for compute. This trade off could be worth it though if we could say that in practice all the inputs tend to have requires grad.

**Comment by soulitzer** [comment]
Oh I believe that alban is suggesting is something slightly different. The grad_input_mask is a quantity computed at runtime when backward is run, so that would not influence how things are saved during forward.

**Comment by nikitaved** [comment]
Ah, I see, I was not aware of that, will change than.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 2]
```diff
@@ -1475,6 +1475,28 @@ Tensor sparse_mask_like_grad(const Tensor& x, const Tensor& gx) {
...
+    const std::array<bool, 3>& grad_input_mask) {
+  if (!grad.defined()) {
+    return std::make_tuple(Tensor{}, Tensor{}, Tensor{});
+  }
+
+  const auto grad_projected = grad.sparse_mask(self);
+  const auto self_requires_grad = grad_input_mask[0];
+  const auto mat1_requires_grad = grad_input_mask[1];
+  const auto mat2_requires_grad = grad_input_mask[2];
+  return std::make_tuple(
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply torch/csrc/autograd/FunctionsManual.cpp:1494]: yes, it should be safe to assume that

**Review by soulitzer — APPROVED** [review]

**Comment by soulitzer** [comment]
Good catch!
